File: Biblioteca-de-Trabajo/3. CODIGO/Prototipo v1.0/src/negocio/ServicioPerfiles.java
package negocio;

import datos.PerfilRepo;
import datos.UsuarioRepo;
import modelos.Perfil;
import modelos.Usuario;
import java.util.List;

public class ServicioPerfiles implements IServicioFiCitas {
    private PerfilRepo perfilRepo = new PerfilRepo();
    private UsuarioRepo usuarioRepo = new UsuarioRepo();
    private GestorEventos gestor;
    private Usuario usuarioActual;
    private Perfil perfilActual;

    public ServicioPerfiles(GestorEventos gestor, Usuario u, Perfil p) {
        this.gestor = gestor;
        this.usuarioActual = u;
        this.perfilActual = p;
    }

    public static class OpResult {
        public boolean exito;
        public String mensaje;
        public OpResult(boolean e, String m) { exito = e; mensaje = m; }
    }

    public List<Perfil> obtenerPerfiles() {
        return perfilRepo.obtenerTodos();
    }

    public OpResult registrarPerfil(String nombre, String descripcion, List<String> permisos) {
        for (Perfil p : perfilRepo.obtenerTodos()) {
            if (p.getNombre().equalsIgnoreCase(nombre)) {
                return new OpResult(false, "Ya existe un perfil con ese nombre.");
            }
        }

        Perfil nuevo = new Perfil(0, nombre, descripcion, "Activo", permisos);
        Perfil creado = perfilRepo.crear(nuevo);
        
        gestor.notificar(String.valueOf(usuarioActual.getId()), perfilActual.getNombre(), "CREAR_PERFIL", "Perfil: " + nombre);
        return new OpResult(true, "Perfil creado exitosamente.");
    }

    public OpResult modificarPerfil(int id, String nombre, String descripcion, String estado, List<String> permisos) {
        Perfil p = perfilRepo.obtenerPorId(id);
        if (p == null) return new OpResult(false, "Perfil no encontrado.");

        if (!p.getNombre().equalsIgnoreCase(nombre)) {
            for (Perfil ext : perfilRepo.obtenerTodos()) {
                if (ext.getId() != id && ext.getNombre().equalsIgnoreCase(nombre)) {
                    return new OpResult(false, "Ya existe otro perfil con ese nombre.");
                }
            }
        }

        if (estado.equals("Inactivo") && p.getEstado().equals("Activo")) {
            if (p.getNombre().equals("Administrador")) {
                long adminsActivos = perfilRepo.obtenerTodos().stream().filter(ext -> ext.getNombre().equals("Administrador") && ext.getEstado().equals("Activo")).count();
                if (adminsActivos <= 1) return new OpResult(false, "No se puede desactivar el único perfil Administrador activo.");
            }

            List<Usuario> usuariosAsignados = usuarioRepo.obtenerPorPerfil(id);
            if (!usuariosAsignados.isEmpty()) {
                return new OpResult(false, "El perfil tiene usuarios asignados. Reasígnelos antes de desactivar.");
            }
        }

        if (p.getNombre().equals("Administrador") && !permisos.contains("GESTION_PERFILES")) {
            return new OpResult(false, "No se puede retirar el permiso de gestionar perfiles al Administrador.");
        }

        p.setNombre(nombre);
        p.setDescripcion(descripcion);
        p.setEstado(estado);
        p.setPermisos(permisos);
        perfilRepo.actualizar(p);

        gestor.notificar(String.valueOf(usuarioActual.getId()), perfilActual.getNombre(), "MODIFICAR_PERFIL", "Perfil ID: " + id);
        return new OpResult(true, "Perfil modificado exitosamente.");
    }
}

File: Biblioteca-de-Trabajo/3. CODIGO/Prototipo v2.0/src/negocio/ServicioPerfiles.js
import { PerfilRepo } from '../datos/PerfilRepo.js';
import { Perfil } from '../modelos/Perfil.js';
import { Evento } from './Evento.js';

export class ServicioPerfiles {
  constructor(gestor, usuario, perfil, perfilRepo) {
    this.gestor = gestor;
    this.usuario = usuario || (typeof window !== 'undefined' ? window.appState?.currentUser : null);
    this.perfil = perfil || (typeof window !== 'undefined' ? window.appState?.currentProfile : null);
    this.perfilRepo = perfilRepo || new PerfilRepo();
  }

  async obtenerPerfiles() {
    return this.perfilRepo.listar();
  }

  async obtenerPerfilPorId(id) {
    if (typeof this.perfilRepo.obtenerPorId !== 'function') return null;
    return this.perfilRepo.obtenerPorId(id);
  }

  normalizarCadena(valor) {
    return typeof valor === 'string' ? valor.trim() : '';
  }

  normalizarPermisos(permisos) {
    if (!Array.isArray(permisos)) return [];
    return permisos.map((permiso) => this.normalizarCadena(permiso)).filter(Boolean);
  }

  esPerfilAdministrador(perfil) {
    return this.normalizarCadena(perfil?.nombre).toLowerCase() === 'administrador';
  }

  async registrarPerfil(datosPerfil) {
    const { nombre, descripcion, estado = 'Activo', permisos = [] } = datosPerfil || {};
    const nombreLimpio = this.normalizarCadena(nombre);
    const descripcionLimpia = this.normalizarCadena(descripcion);
    const permisosLimpios = this.normalizarPermisos(permisos);

    if (!nombreLimpio) {
      return { exito: false, mensaje: 'El nombre del perfil es obligatorio' };
    }
    if (!descripcionLimpia) {
      return { exito: false, mensaje: 'La descripción del perfil es obligatoria' };
    }
    if (permisosLimpios.length === 0) {
      return { exito: false, mensaje: 'Debe seleccionar al menos un permiso' };
    }

    const perfiles = await this.obtenerPerfiles();
    const nombreDuplicado = perfiles.some((perfil) => this.normalizarCadena(perfil.nombre).toLowerCase() === nombreLimpio.toLowerCase());
    if (nombreDuplicado) {
      return { exito: false, mensaje: 'Ya existe un perfil con ese nombre' };
    }

    const perfil = new Perfil({
      id: Date.now(),
      nombre: nombreLimpio,
      descripcion: descripcionLimpia,
      estado,
      permisos: permisosLimpios,
    });

    await this.perfilRepo.guardar(perfil);
    this.notificarEvento('PERFIL_REGISTRADO', null, perfil);
    return { exito: true, mensaje: 'Perfil registrado', perfil };
  }

  async modificarPerfil(id, datosPerfil) {
    const { nombre, descripcion, estado = 'Activo', permisos = [] } = datosPerfil || {};
    const idNumerico = Number(id);
    const nombreLimpio = this.normalizarCadena(nombre);
    const descripcionLimpia = this.normalizarCadena(descripcion);
    const permisosLimpios = this.normalizarPermisos(permisos);

    const perfiles = await this.obtenerPerfiles();
    const perfilActual = perfiles.find((perfil) => Number(perfil.id) === idNumerico);
    if (!perfilActual) {
      return { exito: false, mensaje: 'Perfil no encontrado' };
    }

    if (!nombreLimpio) {
      return { exito: false, mensaje: 'El nombre del perfil es obligatorio' };
    }
    if (!descripcionLimpia) {
      return { exito: false, mensaje: 'La descripción del perfil es obligatoria' };
    }
    if (permisosLimpios.length === 0) {
      return { exito: false, mensaje: 'Debe seleccionar al menos un permiso' };
    }

    const nombreDuplicado = perfiles.some((perfil) => Number(perfil.id) !== idNumerico && this.normalizarCadena(perfil.nombre).toLowerCase() === nombreLimpio.toLowerCase());
    if (nombreDuplicado) {
      return { exito: false, mensaje: 'Ya existe un perfil con ese nombre' };
    }

    if (estado === 'Inactivo' && this.esPerfilAdministrador(perfilActual)) {
      const administradoresActivos = perfiles.filter((perfil) => this.esPerfilAdministrador(perfil) && perfil.estado === 'Activo');
      if (administradoresActivos.length <= 1) {
        return { exito: false, mensaje: 'No se puede desactivar el único perfil administrador activo' };
      }
    }

    const perfilActualizado = new Perfil({
      ...perfilActual,
      nombre: nombreLimpio,
      descripcion: descripcionLimpia,
      estado,
      permisos: permisosLimpios,
    });

    const resultado = await this.perfilRepo.actualizar(idNumerico, perfilActualizado);
    if (!resultado) {
      return { exito: false, mensaje: 'Perfil no encontrado' };
    }

    this.notificarEvento('PERFIL_ACTUALIZADO', perfilActual, perfilActualizado);
    return { exito: true, mensaje: 'Perfil actualizado', perfil: perfilActualizado };
  }

  notificarEvento(tipo, anteriores, modificados) {
    const contexto = {
      usuario: this.usuario || (typeof window !== 'undefined' ? window.appState?.currentUser : null),
      perfil: this.perfil || (typeof window !== 'undefined' ? window.appState?.currentProfile : null),
      anteriores,
      modificados,
    };

    if (!this.gestor) return;
    this.gestor.notificar(new Evento(tipo, contexto));
  }
}
